azurelinuxagent/common/cgroupstelemetry.py

```diff
@@ -13,47 +13,184 @@
+            },
+            "cpu": {
+                "cur_cpu": CGroupsTelemetry._get_metrics_list(current_cpu_usage)
+            }
+        }
+        return processed_extension
 
     @staticmethod
     def metrics_hierarchies():
-        return CGroupsTelemetry._hierarchies
+        pass  # return CGroupsTelemetry._hierarchies
 
     @staticmethod
     def track_cgroup(cgroup):
         """
-        Adds the given item to the list of tracked cgroups
+        Adds the given item to the dictionary of tracked cgroups
         """
-        pass  # TODO
+        # Making track thread-safe.
+        CGroupsTelemetry._rlock.acquire()
+        CGroupsTelemetry._tracked.append(cgroup)
+        # if not CGroupsTelemetry.is_tracked(cgroup.name, cgroup.controller):
+        #     CGroupsTelemetry._tracked[cgroup.name] = {cgroup.controller: cgroup}
+
+        CGroupsTelemetry._rlock.release()
 
     @staticmethod
-    def is_tracked(name):
+    def is_tracked(name, controller):
         """
         Returns true if the given item is in the list of tracked items
         """
-        pass  # TODO
+        return name in CGroupsTelemetry._tracked and controller in CGroupsTelemetry._tracked[name]
 
     @staticmethod
     def stop_tracking(name):
         """
         Stop tracking the cgroups for the given name
         """
-        if CGroupsTelemetry.is_tracked(name):
-            del (CGroupsTelemetry._tracked[name])  # TODO - thread-safeness; review implementation
+        # Pop is atomic and thread-safe.
+        return CGroupsTelemetry._tracked.pop(name, None)
 
     @staticmethod
     def collect_all_tracked():
-        pass  # TODO
+        collected_metrics = {}
+        for name, cgroup_metrics in CGroupsTelemetry._cgroup_metrics.items():
+            collected_metrics[name] = CGroupsTelemetry._process_cgroup_metric(cgroup_metrics)
+            cgroup_metrics.clear()
+
+        return collected_metrics
+
+    @staticmethod
+    def poll_all_tracked():
+        CGroupsTelemetry._rlock.acquire()
+        for cgroup in CGroupsTelemetry._tracked:
+            if cgroup.name not in CGroupsTelemetry._cgroup_metrics:
+                CGroupsTelemetry._cgroup_metrics[cgroup.name] = CgroupMetrics()
+
+            CGroupsTelemetry._cgroup_metrics[cgroup.name].add_new_data(cgroup.controller,
+                                                                       cgroup.collect())
+            if not cgroup.is_active():
+                CGroupsTelemetry._tracked.remove(cgroup)
+
+        CGroupsTelemetry._rlock.release()
+
+    @staticmethod
+    def prune_all_tracked():
+        CGroupsTelemetry._rlock.acquire()
+        for cgroup in CGroupsTelemetry._tracked:
+            if not cgroup.is_active():
+                CGroupsTelemetry._tracked.remove(cgroup)
+
+        CGroupsTelemetry._rlock.release()
+
+    @staticmethod
+    def cleanup():
+        CGroupsTelemetry._tracked *= 0  # emptying the list
```

Comment:
This is a pretty neat way of doing this. I hope it's compatible with all our Python versions!

Reply:
Verified between Py2 and 3. :)